### ISSUE TITLE
Recovery Strategies for Network Parition / Non Automated Destructive Tasks.

### ISSUE THREAD

**Issue by Smithx10** [hashicorp/nomad#5673]
I believe currently in the scenario of a network partition tasks will be destroyed on a Clients return to the cluster.  I was curious if it would be possible to add a key to the job spec so that on recovery these jobs are just left as failed, with the option to manually unfail them after manual intervention.  The key thing here.... is that I'd love for users to be able to take advantage of the api,  but never risk any destructive action to those allocations unless user initiated. 

https://www.nomadproject.io/docs/configuration/server.html#heartbeat_grace
https://www.nomadproject.io/docs/configuration/server.html#min_heartbeat_ttl
https://www.nomadproject.io/docs/configuration/server.html#max_heartbeats_per_second

**Comment by Smithx10** [comment]
In Triton and Other Clouds there is an idea of "Deletion Protection" perhaps that is something that could help mitigate solve this issue.  This could be the key added to the task.

**Comment by schmichael** [comment]
Just to make sure I understand the desired behavior properly, here's the current behavior when a node crashes:

1. Job A is scheduled and Alloc B is placed on Node N
2. Node N crashes
3. After missing enough hearts it is considered `lost`
4. Scheduler creates Alloc C for Job A on Node M.
5. Node N comes back up
6. Node N sees Alloc B was marked as lost and cleans it up

So Job A is running on Node M, but you'd like Node N to reclaim the alloc? Approximately:

6. Node N sees Alloc B was marked as lost and does not start or clean it up yet
7. Operator investigates Node N and runs a `nomad alloc restore B` (made up command)
8. Alloc C is stopped and Alloc B is restarted.

I'm assuming this is useful for jobs like VMs which can have considerable amounts of local state that you'd like to recover whenever possible? If that's the case would we also need to prevent rescheduling `lost` allocs onto new nodes?

**Comment by schmichael** [comment]
Chatted with some team members internally and perhaps what you want is something like task _pinning_ where an allocation is scheduled normally (maybe with a node specific constraint, maybe not), but once placed the job will not be rescheduled on another node until unpinned. If the node becomes lost the alloc will not be rescheduled but will restart instead of terminate if the node comes back.

We have a rough sketch of a feature like this on our internal roadmap, but no estimated release. We'd love to know more details about your use case to help us guide the development of this feature!

**Comment by Smithx10** [comment]
@schmichael,  

You understand the scenario completely.   To clarify, a Restart will attempt to use the same allocation names on the same Client correct?  If so I think I just need to check in StartTask if that instance exists and don't create it if so. 

Thanks Again for your assistance.

**Comment by Smithx10** [comment]
@schmichael 

```
So Job A is running on Node M, but you'd like Node N to reclaim the alloc? Approximately:

Node N sees Alloc B was marked as lost and does not start or clean it up yet
Operator investigates Node N and runs a nomad alloc restore B (made up command)
Alloc C is stopped and Alloc B is restarted.
```

The other option there Operator sees that the Stateful application recovered itself properly Alloc C, and can then run "nomad alloc delete B ( I think this is a made up command )"  because we are healthy and have confirmed that we still have the data that may not have been replicated from Alloc C.

**Comment by Smithx10** [comment]
@schmichael ,

I thought about how the driver could handle this.    Currently when a client fails,  it mark's the allocation lost, and then failed, and then deploys a new allocation id.   If there was an option to try run that that same allocation on a different agent,  we could handle see if triton is already running that allocation and then just mark it up without doing a full creation.  I believe restart currently removes the allocation before creating it again so I believe that should work?

```
Deployed
Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
consul      3        3       1        2          2019-08-06T16:13:03Z

Allocations
ID        Node ID   Task Group  Version  Desired  Status   Created     Modified
6ef43444  3585777e  consul      0        run      running  7m33s ago   6m46s ago
954d08e3  3585777e  consul      0        run      running  7m33s ago   6m45s ago
09c9f081  84c6513b  consul      0        stop     failed   34m36s ago  6m56s ago
2108cf16  84c6513b  consul      0        stop     failed   34m36s ago  6m56s ago
22404ffd  3585777e  consul      0        run      running  34m36s ago  33m50s ago
```

**Comment by schmichael** [comment]
Remote Task Drivers merged in #10248 and released in Nomad 1.1.0 should fix this.

The Remote Task Driver features allow task driver plugins to opt into receiving a new signal `DETACH`. If a task driver sets the `RemoteTasks=true` capability, Nomad will send tasks the `DETACH` signal when the client agent is being drained or starting back up after being down.

The `DETACH` signal indicates that while the allocation is terminal, and the task should no longer be tracked locally, the *remote* task process (eg Triton) is still running and being managed by a new allocation + client node.

See https://www.nomadproject.io/docs/internals/plugins/task-drivers#remote-task-drivers and https://www.nomadproject.io/docs/drivers/remote for details.

While the Remote Task Drivers feature is still considered experimental, I'm closing this issue because it should be addressed! Feel free to reopen or open a new issue if I missed something!
